File: 백준/Silver/2805. 나무 자르기/나무 자르기.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());       // 나무 수  
        int M = Integer.parseInt(st.nextToken());       // 나무 길이  

        st = new StringTokenizer(br.readLine());
        int[] trees = new int[N];
        for (int n = 0; n < N; n++) {
            trees[n] = Integer.parseInt(st.nextToken());
        }
        Arrays.sort(trees);

        // 이진탐색 초기화
        int hi = trees[N-1]+1;
        int lo = 0;

        while (lo < hi) {
            int now = (hi+lo) / 2;
            long sum = 0;
            for (int n=N-1; n>=0; n--){
                if (trees[n] < now) break;
                sum += (trees[n] - now);
            }

            if (sum < M) {
                hi = now;
            } else {
                lo = now + 1;
            }
        }

        System.out.println(lo-1);
    }
}

File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
N, M = map(int, input().split(" "))
trees = list(map(int, input().split(" ")))
trees.sort()
hl = 0
hr = trees[-1]
while hl<hr:
    hm = (hl + hr) // 2
    idx = 0
    for t in range(N):
        if hm < trees[t]:
            idx = t
            break

    m = sum(trees[idx:]) - (hm*(N-idx))
    if m < M: hr = hm
    if m >= M: hl = hm + 1

print(hr-1)